File: app/macos/src/fall_prediction_desktop/ui/i18n.py
from __future__ import annotations

import json
from pathlib import Path


class I18n:
    STORAGE_KEY = "fallguard_lang"

    def __init__(self, locales_dir: Path) -> None:
        self._locales_dir = locales_dir
        self._translations: dict[str, dict] = {}
        self._current_lang: str = "en"
        self._ready: bool = False


    def init(self) -> None:
        self._translations = {}
        for lang in ("en", "zh"):
            path = self._locales_dir / f"{lang}.json"
            try:
                self._translations[lang] = json.loads(path.read_text(encoding="utf-8"))
            except (OSError, json.JSONDecodeError):
                self._translations[lang] = {}

        saved = self._load_saved()
        if saved in self._translations:
            self._current_lang = saved
        self._ready = True

    def t(self, key: str, fallback: str = "") -> str:
        if not self._ready:
            return fallback or key
        value = self._resolve(self._translations.get(self._current_lang, {}), key)
        if value is not None:
            return value
        value = self._resolve(self._translations.get("en", {}), key)
        if value is not None:
            return value
        return fallback or key

    def t_dynamic(self, english_string: str) -> str:
        if not english_string:
            return english_string
        if self._current_lang == "en":
            return english_string
        backend_map = self._translations.get("en", {}).get("backend", {})
        key = backend_map.get(english_string, "")
        if key:
            translated = self.t(key)
            if translated and translated != key:
                return translated
        return english_string

    def set_language(self, lang: str) -> None:
        if lang not in self._translations:
            return
        self._current_lang = lang
        self._save_lang(lang)

    @property
    def language(self) -> str:
        return self._current_lang

    @property
    def ready(self) -> bool:
        return self._ready


    @staticmethod
    def _resolve(lang_obj: dict, path: str) -> str | None:
        parts = path.split(".")
        current = lang_obj
        for part in parts:
            if isinstance(current, dict) and part in current:
                current = current[part]
            else:
                return None
        return current if isinstance(current, str) else None

    def _load_saved(self) -> str:
        try:
            from PySide6.QtCore import QSettings
            settings = QSettings("FallGuard", "FallGuard")
            return settings.value(self.STORAGE_KEY, "en")
        except Exception:
            return "en"

    def _save_lang(self, lang: str) -> None:
        try:
            from PySide6.QtCore import QSettings
            settings = QSettings("FallGuard", "FallGuard")
            settings.setValue(self.STORAGE_KEY, lang)
        except Exception:
            pass


_i18n: I18n | None = None


def get_i18n() -> I18n:
    assert _i18n is not None, "Call init_i18n() first"
    return _i18n


def init_i18n(locales_dir: Path) -> I18n:
    global _i18n
    _i18n = I18n(locales_dir)
    _i18n.init()
    return _i18n


def t(key: str, fallback: str = "") -> str:
    return get_i18n().t(key, fallback)


def t_dynamic(english: str) -> str:
    return get_i18n().t_dynamic(english)

File: app/macos/web/i18n.js
// FallGuard i18n — lightweight translation engine
// Loads en.json + zh.json at init time, switches text via data-i18n + t() calls.

const i18n = (() => {
  const STORAGE_KEY = "fallguard_lang";
  let translations = {};   // merged: { en: {...}, zh: {...} }
  let currentLang = "en";
  let ready = false;

  // Resolve a dotted key like "buttons.startMonitoring" → value
  function resolve(langObj, path) {
    const parts = path.split(".");
    let node = langObj;
    for (const part of parts) {
      if (node == null) return undefined;
      node = node[part];
    }
    return typeof node === "string" ? node : undefined;
  }

  // Public: translate a key to the current language
  function t(key, fallback) {
    if (!ready) return fallback || key;
    const value = resolve(translations[currentLang], key);
    if (value !== undefined) return value;
    // Fallback to English
    const enValue = resolve(translations["en"], key);
    if (enValue !== undefined) return enValue;
    return fallback || key;
  }

  // Public: translate a backend English string (reverse lookup via backend map)
  function tDynamic(englishString) {
    if (!ready || !englishString || currentLang === "en") return englishString;
    // The backend map is a flat dict: English string → i18n dotted key
    const enBackend = translations["en"]?.backend || {};
    const lookupKey = enBackend[englishString];
    if (lookupKey) {
      // lookupKey is something like "status.ready" — resolve it in current language
      return resolve(translations[currentLang], lookupKey) || englishString;
    }
    return englishString;
  }

  // Update all [data-i18n] elements and elements with data-i18n-placeholder
  function updateStaticElements() {
    document.querySelectorAll("[data-i18n]").forEach((el) => {
      const key = el.getAttribute("data-i18n");
      if (key) {
        const translated = t(key, el.textContent);
        if (el.tagName === "TITLE") {
          document.title = translated;
        } else {
          el.textContent = translated;
        }
      }
    });
    document.querySelectorAll("[data-i18n-placeholder]").forEach((el) => {
      const key = el.getAttribute("data-i18n-placeholder");
      if (key) {
        el.placeholder = t(key, el.placeholder);
      }
    });
    document.querySelectorAll("[data-i18n-title]").forEach((el) => {
      const key = el.getAttribute("data-i18n-title");
      if (key) {
        el.title = t(key, el.title);
      }
    });
    document.querySelectorAll("[data-i18n-alt]").forEach((el) => {
      const key = el.getAttribute("data-i18n-alt");
      if (key) {
        el.alt = t(key, el.alt);
      }
    });
  }

  // Apply language: save, update all static text, trigger UI refresh
  function setLanguage(lang) {
    if (lang !== "en" && lang !== "zh") return;
    currentLang = lang;
    try {
      localStorage.setItem(STORAGE_KEY, lang);
    } catch (_) {
      // localStorage unavailable
    }
    updateStaticElements();
    // Update the language select dropdown
    const sel = document.getElementById("languageSelect");
    if (sel) sel.value = lang;
    // Re-render dynamic content by re-polling status
    if (typeof pollStatus === "function") {
      pollStatus();
    }
  }

  // Load JSON files and initialize
  async function init() {
    try {
      const [enResp, zhResp] = await Promise.all([
        fetch("/static/locales/en.json"),
        fetch("/static/locales/zh.json"),
      ]);
      translations["en"] = await enResp.json();
      translations["zh"] = await zhResp.json();
    } catch (_) {
      // If locale files fail to load, translations stays empty and t() returns fallback
      translations["en"] = {};
      translations["zh"] = {};
    }

    // Restore saved language
    let saved = "en";
    try {
      saved = localStorage.getItem(STORAGE_KEY) || "en";
    } catch (_) {
      // localStorage unavailable
    }
    if (saved !== "en" && saved !== "zh") saved = "en";
    currentLang = saved;
    ready = true;

    // Update static elements and language select
    updateStaticElements();
    const sel = document.getElementById("languageSelect");
    if (sel) sel.value = currentLang;
  }

  // Check if i18n is ready
  function isReady() {
    return ready;
  }

  // Get current language
  function getLanguage() {
    return currentLang;
  }

  return { t, tDynamic, setLanguage, init, isReady, getLanguage, updateStaticElements };
})();
